### PR TITLE
Handle duplicate headers in the uri module

### LINKED ISSUE #33325: uri module doesn’t handle multiple cookies with same name
##### ISSUE TYPE

 - Bug Report
 - Feature Idea
 - Documentation Report

##### COMPONENT NAME

    module : uri
##### ANSIBLE VERSION

```
Current master
```

##### STEPS TO REPRODUCE

```yaml
- name: Platform login request
      uri :
        url:  "{{ platform_url }}/login"
        method: POST
        headers:
           Content-Type: "application/json"
        body: "{{ platform_login_request | to_json }}"
        status_code: 200
        validate_certs: no
        timeout: 20
      register: login_result

    - name: Get Pairing Info
      uri :
        url:  "{{ platform_url }}//nodes"
        method: GET
        headers:
           cookie: "{{ login_result.set_cookie}}"
        status_code: 200
        timeout: 20
        validate_certs: no
      
```

##### EXPECTED RESULTS

I'm getting 
set_cookie": “X-JSESSIONID=e80e63ec-defa-44d7-bd6d-b42991d8e1d8; Path=/; Secure; HttpOnly, X-JSESSIONID=deleteMe; Path=/; Max-Age=0; Expires=Sun, 26-Nov-2017 19:35:20 GMT; Secure, X-JSESSIONI=58051d1d-bf80-486c-96ff-2b8fe4dab4ce; Path=/; Secure; HttpOnly, rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Sun, 26-Nov-2017 19:35:21 GMT",

If the server sends the same cookie with same name, the last one should take precedence, but URI module keeps both of them. This behavior is different from that of CURL/python request or urllib2 

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
...and make it easier for users to use cookies by providing a pre-built string

FIxes #33325 
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
uri

##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure unrelated to PR. Restarting CI.

File: lib/ansible/module_utils/urls.py [hunk 1]
```diff
@@ -1049,11 +1049,36 @@ def fetch_url(module, url, data=None, headers=None, method=None,
                      url_password=password, http_agent=http_agent, force_basic_auth=force_basic_auth,
                      follow_redirects=follow_redirects, client_cert=client_cert,
                      client_key=client_key, cookies=cookies)
-        info.update(r.info())
+        # Lowercase keys, to conform to py2 behavior, so that py3 and py2 are predictable
+        info.update(dict((k.lower(), v) for k, v in r.info().items()))
```
Comment (mattclay) [inline lib/ansible/module_utils/urls.py:1056]: This looks like a bug fix. If that's the case it should go into a separate PR, preferably along with tests.

Also, doesn't this cause potential issues with backwards compatibility?

File: lib/ansible/module_utils/urls.py [hunk 2]
```diff
@@ -1049,11 +1049,36 @@ def fetch_url(module, url, data=None, headers=None, method=None,
...
+            # Py3
+            _h = {}
+            for k, v in r.headers.items():
+                # The same as above, lower case keys to match py2 behavior, and create more consistent results
+                _k = k.lower()
+                if _k in _h:
+                    _h[_k] = ', '.join((_h[_k], v))
+                else:
+                    _h[_k] = v
+            info.update(_h)
```
Comment (mattclay) [inline lib/ansible/module_utils/urls.py]: Please use more descriptive variable names for at least `_h` and `_k`, and probably `k` and `v` also. I think it's fine to use some single letter variables in list/dict comprehensions or simple loops, but once there is some additional logic I think readability is improved by using longer names.

Also, it should be fine to reassign the lowercase version of `k` (or whatever it ends up being) to itself since the original value isn't used anywhere else and the type hasn't changed.

File: lib/ansible/module_utils/urls.py [hunk 3]
```diff
@@ -1049,11 +1049,36 @@ def fetch_url(module, url, data=None, headers=None, method=None,
                      url_password=password, http_agent=http_agent, force_basic_auth=force_basic_auth,
                      follow_redirects=follow_redirects, client_cert=client_cert,
                      client_key=client_key, cookies=cookies)
-        info.update(r.info())
+        # Lowercase keys, to conform to py2 behavior, so that py3 and py2 are predictable
+        info.update(dict((k.lower(), v) for k, v in r.info().items()))
+
+        # Don't be lossy, append header values for duplciate headers
```
Comment (mattclay) [inline lib/ansible/module_utils/urls.py]: Fix spelling of `duplicate`.

File: lib/ansible/module_utils/urls.py [hunk 4]
```diff
@@ -1049,11 +1049,36 @@ def fetch_url(module, url, data=None, headers=None, method=None,
...
+            _h = {}
+            for k, v in r.headers.items():
+                # The same as above, lower case keys to match py2 behavior, and create more consistent results
+                _k = k.lower()
+                if _k in _h:
+                    _h[_k] = ', '.join((_h[_k], v))
+                else:
+                    _h[_k] = v
+            info.update(_h)
+        except AttributeError:
```
Comment (mattclay) [inline lib/ansible/module_utils/urls.py]: What causes `AttributeError` to be raised?

**Review by mattclay — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/urls.py (hunk 1 above)
Comment (sivel) [reply lib/ansible/module_utils/urls.py:1056]: I suppose it could be considered a bug fix.

While writing tests I noticed that py2 gives lowercased names, while py3 gives capitalization of the name from the response.

I can't see how it would cause a potential backwards compat issue. I went with the py2 default, because all of our modules should support py2. The only way it could cause a backwards compat issue, is if a module was calling `fetch_url` and only supported python3.  If they supported both py2 and py3, then the code would have to already handle lowercase and mixed case, so it should work as expected.

This normalizes/standardizes return values from this method.

**Comment by sivel** [comment]
@mattclay most comments addressed in a recent commit.  I added a comment to the untouched line around concern of backwards incompat.

File: lib/ansible/module_utils/urls.py (hunk 1 above)
Comment (sivel) [reply lib/ansible/module_utils/urls.py:1056]: Also, in general, this whole PR is a bug fix.  Largely due to difference in behavior of python3 vs python2.  The goal is to normalize the functionality.

The only place not a bugfix is the addition of `cookies_string`.  That was added to ease issues related to the original bug report of missing duplicate headers.  However, due to predictability issues reconstructing the string from the headers, I would consider that part of the bugfix as well.

**Review by mattclay — APPROVED** [review]

**Comment by bersace** [comment]
Hi,

Kudo @sivel. ~~Do you have some input on multiple cookies in `uri` modules ? Unlike what's stated in PR title, the diff seems to modify `get_url` module. Am I wrong ?~~ The integration test playbook is explicit. Thanks.

Regards,
Étienne
